src/Q2/hyowon/Solution.java

```diff
@@ -0,0 +1,21 @@
+package Q2.hyowon;
+
+import java.util.HashMap;
+
+public class Solution {
+    public int solution(String[][] clothes) {
+        HashMap<String, Integer> clothesQuantity = new HashMap<>();
+
+        for (String[] clothe : clothes) {
+            clothesQuantity.put(clothe[1], clothesQuantity.getOrDefault(clothe[1], 0) + 1);
+        }
+
+        int answer = 1;
+        for (String key : clothesQuantity.keySet()) {
```

Comment:
keySet을 사용하신 이유가 있으신가요?

Reply:
이부분 저도 전에 ketset과 entryset performance 차이가 있었던것같아서 궁금해서 찾아보았습니다.
`Map.keyset()` 을 사용해서 `get(key)`해서 key-value를 얻어오는 방법이랑
`Map.entryset()`을 사용해서 `entry.key(), entry.value()`로 ket-value를 얻어오는 방법 중에 
성능차이가 있나 찾아보았는데

https://stackoverflow.com/questions/3870064/performance-considerations-for-keyset-and-entryset-of-map

이부분에서 설명이 잘되어있는것같아요.
결론적으로는HashMap에서 keyset()을 사용한뒤 map.get(key)를 반복문 내에서 한번씩 사용하는동안 내부적으로 연관된 value를 찾기위해 `equals()`와 `hashcode()` 메서드를 사용해서 키를 비교하는 작업이 추가로 들기때문에 entryset()을 사용하는게 더 효율성이 좋다고 하는것 같습니다.


```
Because in the second case, for every key in the keySet the map.get() method is called, which - in the case of a HashMap - requires that the hashCode() and equals() methods of the key object be evaluated in order to find the associated value*. In the first case that extra work is eliminated.
```

Reply:
keySet메소드가 궁금해져서 저도 좀 찾아봤어요
```java
    public Set<K> keySet() {
        Set<K> ks = keySet;
        if (ks == null) {
            ks = new KeySet();
            keySet = ks;
        }
        return ks;
    }
```
위는 keySet 메소드이고, 아래 예제를 실행해 봤어요
```java
Map<String, String> map = new HashMap();
map.put("testKey","testValue");
Set<String> keySet = map.keySet();
```
위 코드에서 
```java
Set<String> keySet = map.keySet();
```
코드가 실행될 때
```java
ks = new KeySet();
```
가 실행되면서 "testKey"를 담은 Set 인스턴스가 리턴되더라구요
그래서 "testKey"를 언제 add해서 리턴시켜주는건지 찾아봤는데 그부분을 못찾겠네요 ㅠㅠ 혹시 아시나요??

Reply:
@JungHyunLyoo 
얼핏 쉬워보여서 찾기 시작했는데 몇시간을 헤매서ㅜㅡㅜ 뭐라도 답을 내고 싶어서 때려맞춰봤습니다.
혹시 잘 아시는 분 계시면 설명을 부탁드립니다!

(`KeySet`이 생성될 때) `HashMap`의 key만 꺼내 Set<>을 만들어주는 소스코드는 따로 없지만
`HashMap`의 inner class인 `KeySet`은 자기 바깥클래스(outer class)의 멤버에 접근 가능하니 
그렇다면 `KeySet`이 생성될 때 바깥클래스를 자유롭게 사용하기 위해 outer class 주소값을 자기 인스턴스 어딘가에 저장해둘 것 같습니다. 

```java
new KeySet(); 
```
이부분에서 생성자가 실행되면서 HashMap<K,V>의 값을 들고갔는데, KeySet에 따로 생성자는 보이지 않고,

```java
// KeySet 클래스 선언부
final class KeySet extends AbstractSet<K> { ... }
// AbstractSet<K> 클래스 생성자
public abstract class AbstractSet<E> extends AbstractCollection<E> implements Set<E> {
    protected AbstractSet() {
    }
    ....
}
```
KeySet의 부모는 AbstractSet<E>이니 부모의 생성자로 생성되면서  K, V 중 K만 남게 된 것이 아닐까요.


이렇게 유추하게 된 까닭은
new KeySet() 이부분 브레이크 포인트 걸고나서 디버깅창(?)에서
```
this = {HashMap@484}   size = 1
ks(slot_1) = null 
```
와 같은 값을 발견하게 되었는데, 
JavaScript에서는 객체 안에 [[prototype]] 라는 internal slot이 있는데, 얘는 상속을 구현하기 위해 자기 부모의 주소값을 저장해두는 공간입니다.
(internal slot은 꼭 상속이 아니더라도 해당 객체가 가지고 있어야 할 어떤 객체의 주소값을 저장해두는 곳)

자바 slot이 자바스크립트의 슬롯과 비슷하다면,,
KeySet은 outer class의 주소값을 들고가서 그걸 사용해 객체를 생성할 것 같다,,
그거 말고 디버깅 창에 보이는 것이 없다,, 그렇게 생각하게 되었습니다.

도움의 손길 기다립니다~